src/auth/providers.js

```diff
@@ -292,7 +298,7 @@ function googleProvider(auth) {
         })
           .then(common.checkStatus)
           .then(response => response.json())
-          .then(json => auth.set(json));
+          .then(json => auth.set(json, providerType));
       }
 
       const redirectUrl = (data && data.redirectUrl) ? data.redirectUrl : undefined;
```

Comment:
[minor] Redirects are not being handled for provider types being prospectively set in advance. You may be able to set a temporary field related to the request like STATE_KEY

Reply:
[major] Speaking of `STATE_KEY`, I'm pretty convinced `window.location.replace(getOAuthLoginURL(auth, 'google', redirectUrl));` and the facebook one are broken with the async changes to storage. getOAuthLoginURL needs to be thenned before using window location replace.